python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -551,6 +555,309 @@ def get_downstream_partitions_for_partitions(
+
+            upstream_dimension_names.remove(upstream_dimension_name)
+            downstream_dimension_names.remove(dimension_mapping.downstream_dimension_name)
+
+        check.invariant(
+            not upstream_dimension_names,
+            "Upstream partitions def has dimensions that are not mapped: {}".format(
+                upstream_dimension_names
+            ),
+        )
+        check.invariant(
+            not downstream_dimension_names,
+            "Downstream partitions def has dimensions that are not mapped: {}".format(
+                downstream_dimension_names
+            ),
+        )
+
+    def _get_dependency_keys(
+        self,
+        a_partitions_def: MultiPartitionsDefinition,
+        a_partition_keys: Sequence[str],
+        b_partitions_def: MultiPartitionsDefinition,
+        a_upstream_of_b: bool,
+    ):
+        """
+        Given two partitions definitions a_partitions_def and b_partitions_def that have a dependency
+        relationship (a_upstream_of_b is True if a_partitions_def is upstream of b_partitions_def),
+        and a_partition_keys, a list of partition keys in a_partitions_def, returns a list of
+        partition keys in the partitions definition b_partitions_def that are
+        dependencies of the partition keys in a_partition_keys.
+        """
+        a_partition_keys_by_dimension = defaultdict(set)
+        for partition_key in a_partition_keys:
+            for dimension_name, key in cast(
+                MultiPartitionKey, partition_key
+            ).keys_by_dimension.items():
+                a_partition_keys_by_dimension[dimension_name].add(key)
+
+        # Maps the dimension name and key of a partition in a_partitions_def to the list of
+        # partition keys in b_partitions_def that are dependencies of that partition
+        dep_b_keys_by_a_dim_and_key: Dict[str, Dict[str, List[str]]] = defaultdict(
+            lambda: defaultdict(list)
+        )
+
+        a_dimension_partitions_defs_by_name = {
+            partitions_def.name: partitions_def.partitions_def
+            for partitions_def in a_partitions_def.partitions_defs
+        }
+        b_dimension_partitions_def_by_name = {
+            partitions_def.name: partitions_def.partitions_def
+            for partitions_def in b_partitions_def.partitions_defs
+        }
+
+        if a_upstream_of_b:
+            # a_partitions_def is upstream of b_partitions_def, so we need to map the
+            # dimension names of a_partitions_def to the corresponding dependent dimensions of
+            # b_partitions_def
+            a_dim_to_dependent_b_dim = {
+                upstream_dim: (
+                    dimension_mapping.downstream_dimension_name,
+                    dimension_mapping,
+                )
+                for upstream_dim, dimension_mapping in self.downstream_mappings_by_upstream_dimension.items()
+            }
+
+            for dimension_name, keys in a_partition_keys_by_dimension.items():
```

Comment:
Nitpick: maybe `dimension_name` -> `a_dim_name` for consistency with `b_dim_name`?